src/audio/smart_amp/smart_amp_maxim_dsm.c

```diff
@@ -0,0 +1,459 @@
+	} else {
+		comp_err(dev,
+			 "[DSM] Feed Forward buffer overflow. (w_ptr : %d + %d > %d)",
+			 *w_ptr, nsamples, DSM_FF_BUF_DB_SZ);
+		return;
+	}
+
+	if (*w_ptr >= DSM_FF_BUF_SZ) {
+		/* Do FF processing */
+		for (x = 0; x < DSM_FRM_SZ; x++) {
+			/* Buffer ordering for DSM : LRLR... -> LL...RR... */
+			if (is_16bit) {
+				input[x] = (buf.buf16[2 * x]);
+				input[x + DSM_FRM_SZ] =
+					(buf.buf16[2 * x + 1]);
+			} else {
+				input[x] = (buf.buf32[2 * x] >> 16);
+				input[x + DSM_FRM_SZ] =
+					(buf.buf32[2 * x + 1] >> 16);
+			}
+		}
+
+		remain = (*w_ptr - DSM_FF_BUF_SZ);
+		if (remain) {
+			if (is_16bit)
+				_memmove(&buf.buf16[0],
+					 &buf.buf16[DSM_FF_BUF_SZ],
+					 remain * szsample);
+			else
+				_memmove(&buf.buf32[0],
+					 &buf.buf32[DSM_FF_BUF_SZ],
+					 remain * szsample);
+		}
+		*w_ptr -= DSM_FF_BUF_SZ;
+
+		/* Do DSM feed forward processing here. Currently bypassed */
+		memcpy_s(output, DSM_FF_BUF_SZ * szsample,
+			 input, DSM_FF_BUF_SZ * szsample);
+
+		for (x = 0; x < DSM_FRM_SZ; x++) {
+			/* Buffer re-ordering LR/LR/LR */
+			if (is_16bit) {
+				buf_out.buf16[*r_ptr + 2 * x] = (output[x]);
+				buf_out.buf16[*r_ptr + 2 * x + 1] =
+					(output[x + DSM_FRM_SZ]);
+			} else {
+				buf_out.buf32[*r_ptr + 2 * x] =
+					(output[x] << 16);
+				buf_out.buf32[*r_ptr + 2 * x + 1] =
+					(output[x + DSM_FRM_SZ] << 16);
+			}
+		}
+
+		*r_ptr += DSM_FF_BUF_SZ;
+	}
+
+	/* Output buffer preparation */
+	if (*r_ptr >= nsamples) {
+		memcpy_s(out, nsamples * szsample,
+			 buf_out.buf16, nsamples * szsample);
+		remain = (*r_ptr - nsamples);
+		if (remain) {
+			if (is_16bit)
+				_memmove(&buf_out.buf16[0],
+					 &buf_out.buf16[nsamples],
+					 remain * szsample);
+			else
+				_memmove(&buf_out.buf32[0],
+					 &buf_out.buf32[nsamples],
+					 remain * szsample);
+		}
+		*r_ptr -= nsamples;
+	} else {
+		memset(out, 0, nsamples * szsample);
+		comp_err(dev,
+			 "[DSM] DSM FF process underrun. r_ptr : %d",
+			 *r_ptr);
+	}
+}
+
+static void maxim_dsm_fb_proc(struct smart_amp_mod_struct_t *hspk,
+			      struct comp_dev *dev, void *in,
+			      int nsamples, int szsample)
+{
+	union smart_amp_buf buf;
+	int *w_ptr = &hspk->buf.fb.avail;
+	int16_t *v = hspk->buf.voltage;
+	int16_t *i = hspk->buf.current;
+	bool is_16bit = szsample == 2 ? true : false;
+	int remain;
+	int x;
+
+	buf.buf16 = (int16_t *)hspk->buf.fb.buf;
+	buf.buf32 = (int32_t *)hspk->buf.fb.buf;
+
+	if (*w_ptr + nsamples <= DSM_FB_BUF_DB_SZ) {
+		if (is_16bit)
+			memcpy_s(&buf.buf16[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		else
+			memcpy_s(&buf.buf32[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		*w_ptr += nsamples;
+	} else {
+		comp_err(dev, "[DSM] Feedback buffer overflow. w_ptr : %d",
+			 *w_ptr);
+		return;
+	}
+	if (*w_ptr >= DSM_FB_BUF_SZ) {
+		/* Do FB processing */
+		for (x = 0; x < DSM_FRM_SZ; x++) {
+			/* Buffer ordering for DSM : IVIV... -> VV... II...*/
+			if (is_16bit) {
```

Comment:
The type check should be outside the for loops.

Reply:
Agreed. Shall fix it.